src/Microsoft.AspNet.Routing/InlineRouteParameterParser.cs

```diff
@@ -12,82 +9,230 @@ namespace Microsoft.AspNet.Routing
+                                currentIndex--;
+                                break;
+                        }
+                        break;
+                    case ParseState.InsideParantheses:
+                        switch (currentChar)
+                        {
+                            case char.MinValue:
+                                state = ParseState.End;
+                                var constraintText = routeParameter.Substring(startIndex, currentIndex - startIndex);
+                                inlineConstraints.Add(new InlineConstraint(constraintText));
+                                break;
+                            case ')':
+                                // Only consume a ')' token if
+                                // (a) it is the last token
+                                // (b) the next character is the start of the new constraint ':'
+                                // (c) the next character is the start of the default parameter.
+
+                                var nextChar = currentIndex + 1 > endIndex ? char.MinValue : routeParameter[currentIndex + 1];
+                                switch (nextChar)
+                                {
+                                    case char.MinValue:
+                                        state = ParseState.End;
+                                        constraintText = routeParameter.Substring(startIndex, currentIndex - startIndex + 1);
+                                        inlineConstraints.Add(new InlineConstraint(constraintText));
+                                        break;
+                                    case ':':
+                                        state = ParseState.Start;
+                                        constraintText = routeParameter.Substring(startIndex, currentIndex - startIndex + 1);
+                                        inlineConstraints.Add(new InlineConstraint(constraintText));
+                                        startIndex = currentIndex + 1;
+                                        break;
+                                    case '=':
+                                        state = ParseState.End;
+                                        constraintText = routeParameter.Substring(startIndex, currentIndex - startIndex + 1);
+                                        inlineConstraints.Add(new InlineConstraint(constraintText));
+                                        break;
+                                }
+                                break;
+                            case ':':
+                            case '=':
+                                // In the original implementation, the Regex would've backtracked if it encountered an
+                                // unbalanced opening bracket followed by (not necessarily immediatiely) a delimiter.
+                                // Simply verifying that the parantheses will eventually be balanced should suffice to
```

Comment:
It's not required to be balanced, just closed.

This is totally valid: `{foo:regex(\\(\\(\\(\\()}`